sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -48,6 +48,13 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
     s"$nodeName $joinType ${buildSide} ($opId)".trim
   }
 
+  /**
+   * When overriding metrics be sure to include numOutputRows & numMatchedRows
+   */
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numMatchedRows" -> SQLMetrics.createMetric(sparkContext, "number of matched rows"))
```

Comment:
I feel it's kind of weird that we need to double define `metrics` here in `HashJoin`, and also its subclass `BroadcastHashJoinExec` and `ShuffledHashJoinExec`. It might introduce unexpected bug later. Could we pass the `numMatchedRows` to `HashJoin`, similar to what we did for `numOutputRows` before?

Reply:
But do we need to define `metrics` here at all? If `numMatchedRows` becomes `lazy` as it should be, then probably we don't need to have `metrics` here.

Reply:
I removed the metrics from HashJoin in the latest commit and made the numMatchedRows lazy.